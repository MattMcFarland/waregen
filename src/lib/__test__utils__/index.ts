import { clone, range } from "@@/utils";
import { DefaultWareEntity, DefaultBlueprintEntity } from "@@/entities/Ware";
import { AddwareEntity } from "@@/XMLTypes/X4WareGenXML";
import { GeneratorConfig } from "@@/generator";

export const mockGeneratorConfig = (): GeneratorConfig => ({
  defaultWare: defaultWare(),
  defaultBlueprint: defaultBlueprint(),
  addwaresList: mockAddwares(4),
  gamePath: "/test/Game",
  modPath: "extensions/test",
  modPrefix: "pfx",
  unpackedPath: "unpacked"
});

export const mockAddwares = (count: number): AddwareEntity[] => {
  const char = (code: number) => "abcdefghijklmnopqrstuvwxyz"[code];

  return range(count).map(
    (i: number): AddwareEntity => ({
      Attributes: {
        baskets: `[test_basket_${char(i)}]`,
        cloneProductionModuleFrom: `prod_cloned_macro_${char(i)}`,
        id: `${char(i)}`
      },
      ware: [mockWare(`${char(i)}`)],
      blueprint: [mockBlueprint(`${char(i)}`)]
    })
  );
};

export const mockWare = (name: string) => {
  const ware = defaultWare();
  ware.Attributes.name = name;
  return clone(ware);
};
export const mockBlueprint = (name: string) => {
  const bp = defaultBlueprint();
  bp.Attributes.name = `${name} production`;
  return clone(bp);
};

export const defaultBlueprint = (): DefaultBlueprintEntity => {
  const bp = <DefaultBlueprintEntity>defaultWare();
  bp.Attributes.tags = "module";
  bp.Attributes.description = "blueprint description";
  bp.restriction = [
    {
      Attributes: { licence: "station_gen_basic " }
    }
  ];
  bp.research = [
    {
      Attributes: { time: 10 },
      research: [
        { ware: [{ Attributes: { ware: "research_module_production" } }] }
      ]
    }
  ];
  bp.owner = [
    { Attributes: { faction: "antigone" } },
    { Attributes: { faction: "argon" } },
    { Attributes: { faction: "holyorder" } },
    { Attributes: { faction: "paranid" } },
    { Attributes: { faction: "teladi" } }
  ];
  bp.production = [
    {
      Attributes: {
        time: 791,
        amount: 1,
        method: "default",
        name: "{20206,101}"
      },
      primary: [
        {
          ware: [
            { Attributes: { ware: "claytronics", amount: 30 } },
            { Attributes: { ware: "energycells", amount: 75 } },
            { Attributes: { ware: "hullparts", amount: 150 } }
          ]
        }
      ]
    }
  ];
  return clone(bp);
};

export const defaultWare = (): DefaultWareEntity =>
  clone({
    Attributes: {
      id: "default",
      name: "default",
      transport: "container",
      volume: 1
    },
    price: [
      {
        Attributes: {
          min: 1,
          average: 1,
          max: 1
        }
      }
    ],
    production: [
      {
        Attributes: {
          time: 10,
          amount: 1,
          method: "default",
          name: "{20206, 101}"
        },
        primary: [
          {
            ware: [{ Attributes: { ware: "energycells", amount: 50 } }]
          }
        ],
        effects: [
          {
            effect: [
              {
                Attributes: { type: "efficiency", product: 1 }
              }
            ]
          }
        ]
      }
    ]
  });
