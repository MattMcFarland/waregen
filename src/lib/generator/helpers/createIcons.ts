import { IconCollection, Icon } from "../../entities";
import pathBuilder from "../../utils/PathBuilder";
import { GeneratorConfig } from "../../generator";
import { IdRoster } from "../../utils/IdRoster";

export default function createIcons(config: GeneratorConfig): IconCollection {
  const icons = config.addwaresList.map(ware => {
    const ids = new IdRoster(ware.Attributes.id, config.modPrefix);
    return new Icon({
      name: ids.iconName,
      texture: pathBuilder(config, ware.Attributes.id)
        .dirModPath()
        .fileIconDDS()
        .toString(),
      width: 256,
      height: 256
    });
  });

  return new IconCollection(icons);
}
