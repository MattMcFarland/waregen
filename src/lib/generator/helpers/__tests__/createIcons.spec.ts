import { mockGeneratorConfig } from "@@/__test__utils__";
import createIcons from "../createIcons";

describe("generator/helpers", () => {
  describe("createIcons", () => {
    test("toJson()", () => {
      const mockConfig = mockGeneratorConfig();
      const macroIndex = createIcons(mockConfig);
      expect(macroIndex.toJson()).toMatchSnapshot();
    });
    test("toXml()", () => {
      const mockConfig = mockGeneratorConfig();
      const macroIndex = createIcons(mockConfig);
      expect(macroIndex.toXml()).toMatchSnapshot();
    });
  });
});
