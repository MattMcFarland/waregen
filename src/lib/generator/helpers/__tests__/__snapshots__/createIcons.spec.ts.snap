// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator/helpers createIcons toJson() 1`] = `"{\\"icon\\":[{\\"Attributes\\":{\\"name\\":\\"module_pfx_prod_gen_ware0_macro\\",\\"texture\\":\\"extensions\\\\\\\\test\\\\\\\\assets\\\\\\\\fx\\\\\\\\gui\\\\\\\\textures\\\\\\\\pfx_prod_gen_ware0_macro.dds\\",\\"width\\":256,\\"height\\":256}},{\\"Attributes\\":{\\"name\\":\\"module_pfx_prod_gen_ware1_macro\\",\\"texture\\":\\"extensions\\\\\\\\test\\\\\\\\assets\\\\\\\\fx\\\\\\\\gui\\\\\\\\textures\\\\\\\\pfx_prod_gen_ware1_macro.dds\\",\\"width\\":256,\\"height\\":256}},{\\"Attributes\\":{\\"name\\":\\"module_pfx_prod_gen_ware2_macro\\",\\"texture\\":\\"extensions\\\\\\\\test\\\\\\\\assets\\\\\\\\fx\\\\\\\\gui\\\\\\\\textures\\\\\\\\pfx_prod_gen_ware2_macro.dds\\",\\"width\\":256,\\"height\\":256}},{\\"Attributes\\":{\\"name\\":\\"module_pfx_prod_gen_ware3_macro\\",\\"texture\\":\\"extensions\\\\\\\\test\\\\\\\\assets\\\\\\\\fx\\\\\\\\gui\\\\\\\\textures\\\\\\\\pfx_prod_gen_ware3_macro.dds\\",\\"width\\":256,\\"height\\":256}},{\\"Attributes\\":{\\"name\\":\\"module_pfx_prod_gen_ware4_macro\\",\\"texture\\":\\"extensions\\\\\\\\test\\\\\\\\assets\\\\\\\\fx\\\\\\\\gui\\\\\\\\textures\\\\\\\\pfx_prod_gen_ware4_macro.dds\\",\\"width\\":256,\\"height\\":256}}]}"`;

exports[`generator/helpers createIcons toXml() 1`] = `
"<icons>
  <icon name=\\"module_pfx_prod_gen_ware0_macro\\" texture=\\"extensions\\\\test\\\\assets\\\\fx\\\\gui\\\\textures\\\\pfx_prod_gen_ware0_macro.dds\\" width=\\"256\\" height=\\"256\\"/>
  <icon name=\\"module_pfx_prod_gen_ware1_macro\\" texture=\\"extensions\\\\test\\\\assets\\\\fx\\\\gui\\\\textures\\\\pfx_prod_gen_ware1_macro.dds\\" width=\\"256\\" height=\\"256\\"/>
  <icon name=\\"module_pfx_prod_gen_ware2_macro\\" texture=\\"extensions\\\\test\\\\assets\\\\fx\\\\gui\\\\textures\\\\pfx_prod_gen_ware2_macro.dds\\" width=\\"256\\" height=\\"256\\"/>
  <icon name=\\"module_pfx_prod_gen_ware3_macro\\" texture=\\"extensions\\\\test\\\\assets\\\\fx\\\\gui\\\\textures\\\\pfx_prod_gen_ware3_macro.dds\\" width=\\"256\\" height=\\"256\\"/>
  <icon name=\\"module_pfx_prod_gen_ware4_macro\\" texture=\\"extensions\\\\test\\\\assets\\\\fx\\\\gui\\\\textures\\\\pfx_prod_gen_ware4_macro.dds\\" width=\\"256\\" height=\\"256\\"/>
</icons>"
`;
